File: drawlogic/web/js/geometry.js
// Affine geometry. A direct port of drawlogic/geometry.py -- the canvas and
// the exporter must place a pin in exactly the same spot.

export function fmt(value, places = 3) {
  let text = Number(value).toFixed(places);
  if (text.indexOf(".") >= 0) {
    text = text.replace(/0+$/, "").replace(/\.$/, "");
  }
  if (text === "" || text === "-" || text === "-0") return "0";
  return text;
}

export class Affine {
  constructor(a = 1, b = 0, c = 0, d = 1, e = 0, f = 0) {
    this.a = a; this.b = b; this.c = c;
    this.d = d; this.e = e; this.f = f;
  }

  static translate(tx, ty) {
    return new Affine(1, 0, 0, 1, tx, ty);
  }

  static scale(sx, sy) {
    if (sy === undefined) sy = sx;
    return new Affine(sx, 0, 0, sy, 0, 0);
  }

  static rotate(degrees) {
    const rad = (degrees * Math.PI) / 180;
    const cos = Math.cos(rad);
    const sin = Math.sin(rad);
    return new Affine(cos, sin, -sin, cos, 0, 0);
  }

  // Returns the transform that applies `other` first, then this one.
  multiply(other) {
    return new Affine(
      this.a * other.a + this.c * other.b,
      this.b * other.a + this.d * other.b,
      this.a * other.c + this.c * other.d,
      this.b * other.c + this.d * other.d,
      this.a * other.e + this.c * other.f + this.e,
      this.b * other.e + this.d * other.f + this.f
    );
  }

  apply(x, y) {
    return [this.a * x + this.c * y + this.e, this.b * x + this.d * y + this.f];
  }

  scaleFactor() {
    const sx = Math.hypot(this.a, this.b);
    const sy = Math.hypot(this.c, this.d);
    return (sx + sy) / 2;
  }

  toSvg() {
    return `matrix(${[this.a, this.b, this.c, this.d, this.e, this.f]
      .map((v) => fmt(v, 4))
      .join(",")})`;
  }
}

// `x, y` is the top-left of the unrotated box, and rotation happens about its
// centre, so rotating a cell never moves it.
export function cellMatrix(x, y, w, h, sw, sh, rotate = 0, mirror = false) {
  const cx = x + w / 2;
  const cy = y + h / 2;
  let sx = w / sw;
  const sy = h / sh;
  if (mirror) sx = -sx;

  let m = Affine.translate(cx, cy);
  m = m.multiply(Affine.rotate(rotate));
  m = m.multiply(Affine.scale(sx, sy));
  m = m.multiply(Affine.translate(-sw / 2, -sh / 2));
  return m;
}

export function corners(x, y, w, h) {
  return [[x, y], [x + w, y], [x + w, y + h], [x, y + h]];
}

export function boundsOf(points) {
  if (!points.length) return null;
  const xs = points.map((p) => p[0]);
  const ys = points.map((p) => p[1]);
  const x0 = Math.min(...xs);
  const y0 = Math.min(...ys);
  return [x0, y0, Math.max(...xs) - x0, Math.max(...ys) - y0];
}

// ---- the symbol library ----

let library = {};

export function setLibrary(data) {
  library = data || {};
}

// Blocks standing in for referenced drawings arrive with the drawing that
// references them, not with the library, so they are merged in on open. The
// previous drawing's are dropped first: the same ref means a different file
// from a different folder, and a stale one would draw the wrong pins.
export function setSheets(data) {
  for (const id of Object.keys(library)) {
    if (id.startsWith("sheet:")) delete library[id];
  }
  Object.assign(library, data || {});
}

export function get(typeId) {
  return library[typeId] || null;
}

// The symbol a placed cell draws with. A cell that references another drawing
// takes its symbol from that drawing's ports, so the lookup is by ref rather
// than by type; null when the reference has not been resolved, which reads the
// same as an unknown type.
export function forCell(cell) {
  if (!cell) return null;
  if (cell.ref) return library[`sheet:${cell.ref}`] || null;
  return library[cell.type] || null;
}

export function ids() {
  return Object.keys(library).sort();
}

// The palette: only symbols you can pick up and place. A block standing in
// for another drawing is a real symbol to everything that draws or routes, but
// it comes from that drawing's ports rather than from the library, so there is
// nothing to offer.
export function byCategory() {
  const groups = {};
  for (const id of ids()) {
    if (library[id].listed === false) continue;
    const category = library[id].category || "misc";
    (groups[category] = groups[category] || []).push(id);
  }
  return groups;
}

export function findPin(symbol, name) {
  if (!symbol) return null;
  return symbol.pins.find((pin) => pin.name === name) || null;
}

// `scale` is the document-wide symbol scale. It grows a cell about its own
// centre, so turning every gate up does not drag the layout sideways.
export function matrixFor(symbol, cell, scale = 1) {
  let x = cell.x || 0;
  let y = cell.y || 0;
  let w = cell.w === undefined ? symbol.size[0] : cell.w;
  let h = cell.h === undefined ? symbol.size[1] : cell.h;

  if (scale !== 1) {
    const cx = x + w / 2;
    const cy = y + h / 2;
    w *= scale;
    h *= scale;
    x = cx - w / 2;
    y = cy - h / 2;
  }

  return cellMatrix(x, y, w, h, symbol.size[0], symbol.size[1],
                    cell.rotate || 0, Boolean(cell.mirror));
}

export function pinPosition(symbol, cell, pinName, scale = 1) {
  const pin = findPin(symbol, pinName);
  if (!pin) return null;
  return matrixFor(symbol, cell, scale).apply(pin.x, pin.y);
}

export function cellBounds(symbol, cell, scale = 1) {
  const matrix = matrixFor(symbol, cell, scale);
  const points = corners(0, 0, symbol.size[0], symbol.size[1])
    .map(([px, py]) => matrix.apply(px, py));
  return boundsOf(points);
}

File: drawlogic/geometry.py
"""Affine geometry for placing symbols and resolving pin locations.

Drawn shapes and pin coordinates are pushed through the same matrix, so a
pin can never resolve to a place the symbol is not actually drawn.

Usage:

    from drawlogic.geometry import Affine, cell_matrix, fmt

    m = cell_matrix(x=100, y=50, w=60, h=40,      # where the cell sits
                    sw=60, sh=40,                 # the symbol's natural size
                    rotate=90, mirror=False)
    m.apply(60, 20)      # a symbol-local point in sheet coordinates
    m.to_svg()           # "matrix(a,b,c,d,e,f)" for an SVG transform
    fmt(3.140000)        # "3.14" -- trims trailing zeros for smaller output
"""

import math


def fmt(value, places=3):
  """Format a number for SVG output without trailing zeros."""
  text = ("%.*f" % (places, float(value))).rstrip("0").rstrip(".")
  if text in ("", "-", "-0"):
    return "0"
  return text


class Affine(object):
  """A 2D affine transform, stored as the SVG matrix(a, b, c, d, e, f).

      | a  c  e |
      | b  d  f |
  """

  __slots__ = ("a", "b", "c", "d", "e", "f")

  def __init__(self, a=1.0, b=0.0, c=0.0, d=1.0, e=0.0, f=0.0):
    self.a = float(a)
    self.b = float(b)
    self.c = float(c)
    self.d = float(d)
    self.e = float(e)
    self.f = float(f)

  @staticmethod
  def identity():
    return Affine()

  @staticmethod
  def translate(tx, ty):
    return Affine(1, 0, 0, 1, tx, ty)

  @staticmethod
  def scale(sx, sy=None):
    if sy is None:
      sy = sx
    return Affine(sx, 0, 0, sy, 0, 0)

  @staticmethod
  def rotate(degrees):
    rad = math.radians(degrees)
    cos = math.cos(rad)
    sin = math.sin(rad)
    return Affine(cos, sin, -sin, cos, 0, 0)

  def multiply(self, other):
    """Return the transform that applies `other` first, then `self`."""
    return Affine(
      self.a * other.a + self.c * other.b,
      self.b * other.a + self.d * other.b,
      self.a * other.c + self.c * other.d,
      self.b * other.c + self.d * other.d,
      self.a * other.e + self.c * other.f + self.e,
      self.b * other.e + self.d * other.f + self.f,
    )

  def apply(self, x, y):
    return (self.a * x + self.c * y + self.e,
            self.b * x + self.d * y + self.f)

  def scale_factor(self):
    """Average linear scale, used to keep stroke widths visually stable."""
    sx = math.hypot(self.a, self.b)
    sy = math.hypot(self.c, self.d)
    return (sx + sy) / 2.0

  def to_svg(self):
    parts = [fmt(v, 4) for v in (self.a, self.b, self.c, self.d, self.e, self.f)]
    return "matrix(%s)" % ",".join(parts)

  def is_identity(self):
    return (self.a == 1.0 and self.b == 0.0 and self.c == 0.0
            and self.d == 1.0 and self.e == 0.0 and self.f == 0.0)

  def __repr__(self):
    return "Affine(%s)" % ", ".join(
      fmt(v, 4) for v in (self.a, self.b, self.c, self.d, self.e, self.f))


def cell_matrix(x, y, w, h, sw, sh, rotate=0, mirror=False):
  """Build the transform placing a symbol into sheet coordinates.

  The symbol's natural box is (sw, sh); it is scaled to (w, h), optionally
  mirrored left-to-right, then rotated about the centre of that box. `x, y`
  is the top-left of the unrotated box, so rotating a cell never moves its
  centre -- which is what makes repeated rotation feel predictable.
  """
  if sw <= 0 or sh <= 0:
    raise ValueError("symbol size must be positive, got %r x %r" % (sw, sh))

  cx = x + w / 2.0
  cy = y + h / 2.0
  sx = w / float(sw)
  sy = h / float(sh)
  if mirror:
    sx = -sx

  m = Affine.translate(cx, cy)
  m = m.multiply(Affine.rotate(rotate))
  m = m.multiply(Affine.scale(sx, sy))
  m = m.multiply(Affine.translate(-sw / 2.0, -sh / 2.0))
  return m


def bbox_of_points(points):
  """Return (x, y, w, h) covering the given (x, y) pairs, or None if empty."""
  if not points:
    return None
  xs = [p[0] for p in points]
  ys = [p[1] for p in points]
  return (min(xs), min(ys), max(xs) - min(xs), max(ys) - min(ys))


def union_bbox(a, b):
  if a is None:
    return b
  if b is None:
    return a
  x0 = min(a[0], b[0])
  y0 = min(a[1], b[1])
  x1 = max(a[0] + a[2], b[0] + b[2])
  y1 = max(a[1] + a[3], b[1] + b[3])
  return (x0, y0, x1 - x0, y1 - y0)


def expand_bbox(box, margin):
  if box is None:
    return None
  return (box[0] - margin, box[1] - margin,
          box[2] + 2 * margin, box[3] + 2 * margin)


def corners(x, y, w, h):
  return [(x, y), (x + w, y), (x + w, y + h), (x, y + h)]


def snap(value, step):
  if not step:
    return value
  return round(value / float(step)) * step
